### PR TITLE
Add query_ellipse functionality.

### PR DESCRIPTION
**Pull request by erykoff**
Hey @wmwv ! This adds `query_ellipse` functionality.  Doing the tests were a bit of a challenge, because it's just comparing my c code to some python code.  I don't know if you have any better ideas to confirm that the pixels are in an ellipse.

But this bit of code (requiring `skyproj` and `healsparse`) shows that things are working as expected:

```
import numpy as np
import matplotlib.pyplot as plt
import hpgeom
import healsparse as hsp
import skyproj


nside = 4096
hspmap = hsp.HealSparseMap.make_empty(32, nside, np.int32)

lon = 90.0
lat = 20.0
a = 3.0
b = 2.0
alpha = 45.0

pix = hpgeom.query_ellipse(nside, lon, lat, a, b, alpha)
hspmap[pix] = 1

fig = plt.figure(1, figsize=(8, 5))
fig.clf()
ax = fig.add_subplot(111)
sp = skyproj.McBrydeSkyproj(ax=ax, lon_0=lon)
sp.draw_hspmap(hspmap, lon_range=[85, 95], lat_range=[15, 25])
sp.ellipse(lon, lat, a, b, alpha, color='red', linewidth=2)
plt.show()
```

### CODE REVIEW

File: tests/test_query_ellipse.py [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+    vec = hpgeom.angle_to_vector(lon, lat)
+
+    a = np.deg2rad(major)
+    b = np.deg2rad(minor)
+
+    gamma = np.sqrt(a**2. - b**2.)
+    alpha = np.deg2rad(alpha_deg)
+
+    F1_x = (np.cos(alpha)*np.sin(gamma)*np.cos(phi_0)*np.cos(theta_0)
+            + np.sin(alpha)*np.sin(gamma)*np.sin(phi_0)
```
Comment (wmwv) [inline tests/test_query_ellipse.py:100]:
```suggestion
- np.sin(alpha)*np.sin(gamma)*np.sin(phi_0)
```

File: tests/test_query_ellipse.py [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+            + np.cos(gamma)*np.cos(phi_0)*np.sin(theta_0))
+    F2_x = (-np.cos(alpha)*np.sin(gamma)*np.cos(phi_0)*np.cos(theta_0)
+            - np.sin(alpha)*np.sin(gamma)*np.sin(phi_0)
```
Comment (wmwv) [inline tests/test_query_ellipse.py:103]:
```suggestion
            + np.sin(alpha)*np.sin(gamma)*np.sin(phi_0)
```

File: tests/test_query_ellipse.py [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+    theta, phi = hpgeom.lonlat_to_thetaphi(lon, lat)
+    theta_0, phi_0 = hpgeom.lonlat_to_thetaphi(lon_0, lat_0)
+    vec = hpgeom.angle_to_vector(lon, lat)
+
+    a = np.deg2rad(major)
+    b = np.deg2rad(minor)
+
+    gamma = np.sqrt(a**2. - b**2.)
+    alpha = np.deg2rad(alpha_deg)
+
```
Comment (wmwv) [inline tests/test_query_ellipse.py:98]:
```suggestion
    # The terms for _x and _y are different in their sign flipping
    # F1_x: + - +; F2_x: - + +;
    # F1_y: + + +; F2_y: - - +;
```

File: hpgeom/healpix_geom.c [hunk 4]
```diff
@@ -1302,3 +1302,108 @@ void query_polygon(healpix_info *hpx, pointingarr *vertex, int fact, i64rangeset
...
+    double sin_alpha = sin(alpha);
+    double gamma = sqrt(semi_major * semi_major - semi_minor * semi_minor);
+    double sin_gamma = sin(gamma);
+    double cos_gamma = cos(gamma);
+    double cos_phi = cos(ptg_phi);
+    double cos_theta = cos(ptg_theta);
+    double sin_phi = sin(ptg_phi);
+    double sin_theta = sin(ptg_theta);
+
+    f1vec.x = cos_alpha * sin_gamma * cos_phi * cos_theta + sin_alpha * sin_gamma * sin_phi +
```
Comment (wmwv) [inline hpgeom/healpix_geom.c:1346]:
```suggestion
f1vec.x = cos_alpha * sin_gamma * cos_phi * cos_theta - sin_alpha * sin_gamma * sin_phi +
```

File: hpgeom/healpix_geom.c [hunk 5]
```diff
@@ -1302,3 +1302,108 @@ void query_polygon(healpix_info *hpx, pointingarr *vertex, int fact, i64rangeset
...
+              cos_gamma * cos_phi * sin_theta;
+    f2vec.x = -cos_alpha * sin_gamma * cos_phi * cos_theta - sin_alpha * sin_gamma * sin_phi +
```
Comment (wmwv) [inline hpgeom/healpix_geom.c:1348]:
```suggestion
-cos_alpha * sin_gamma * cos_phi * cos_theta + sin_alpha * sin_gamma * sin_phi +
```

**Comment by wmwv** [comment]
Nice way to go with this.

The current code rotates ellipses the wrong way for lon!=90 (270).

I think it's because there's a sign error in the calculation of F1,2_x.  The terms should be +-+, -++.

```
import numpy as np
import matplotlib.pyplot as plt
import hpgeom
import healsparse as hsp
import skyproj


nside = 4096
hspmap = hsp.HealSparseMap.make_empty(32, nside, np.int32)

lon = 0.0
lat = 20.0
a = 3.0
b = 2.0
alpha = 45.0

pix = hpgeom.query_ellipse(nside, lon, lat, a, b, alpha)
hspmap[pix] = 1

fig = plt.figure(1, figsize=(8, 5))
fig.clf()
ax = fig.add_subplot(111)
sp = skyproj.McBrydeSkyproj(ax=ax)
sp.draw_hspmap(hspmap)
sp.ellipse(lon, lat, a, b, alpha, color='red', linewidth=2, nsamp=1000)
plt.show()
```

gives

ellipse_0_20_3_2_45

File: hpgeom/healpix_geom.c [hunk 6]
```diff
@@ -1302,3 +1302,108 @@ void query_polygon(healpix_info *hpx, pointingarr *vertex, int fact, i64rangeset
...
+
+    /*
+      The following math is adapted from
+      https://math.stackexchange.com/questions/3747965/points-within-an-ellipse-on-the-globe
+
+      This foci of the ellipse are pre-computed from the center, semi-major and semi-minor
+      axes, and the rotation angle (alpha).  This is a lot of trig, but it only has to
+      be done once per query and not per pixel.
+
+      The metric is then for pixels that the sum of the distance between the two foci
```
Comment (wmwv) [inline hpgeom/healpix_geom.c]: Reword to make more clear.  You don't mean the "distance between the two foci"

Something like:
"The criterion is then that the sum of the distances from a pixel to each of the foci sum to less than 2*semi_major."

**Review by wmwv — CHANGES_REQUESTED** [review]
Rotates ellipses the wrong way for lon away from 90 degrees.

**Comment by wmwv** [comment]
Ah, I think I understand.  In some sense my suggestion to fix is exactly backward.

To go from alpha being measured West of North (what I assume the StackOverflow answer meant by CCW from North) to measured East of North, you flipped the sin(alpha) terms in the first lines for F1,2_x, but forgot to do so in the second line for F1,2_y.

It should be

```
    # The terms for _x and _y are different in their sign flipping
    # F1_x: + + +; F2_x: - - +;
    # The sin(alpha) terms are flipped w.r.t. to the StackExchange post
    # because we're using the astronomy position angle convention of
    # alpha being measured East of North.
    F1_x = (np.cos(alpha)*np.sin(gamma)*np.cos(phi_0)*np.cos(theta_0)
            + np.sin(alpha)*np.sin(gamma)*np.sin(phi_0)
            + np.cos(gamma)*np.cos(phi_0)*np.sin(theta_0))
    F2_x = (-np.cos(alpha)*np.sin(gamma)*np.cos(phi_0)*np.cos(theta_0)
            - np.sin(alpha)*np.sin(gamma)*np.sin(phi_0)
            + np.cos(gamma)*np.cos(phi_0)*np.sin(theta_0))
    # F1_y: + - +; F2_y: - + +;
    F1_y = (np.cos(alpha)*np.sin(gamma)*np.sin(phi_0)*np.cos(theta_0)
            - np.sin(alpha)*np.sin(gamma)*np.cos(phi_0)
            + np.cos(gamma)*np.sin(phi_0)*np.sin(theta_0))
    F2_y = (-np.cos(alpha)*np.sin(gamma)*np.sin(phi_0)*np.cos(theta_0)
            + np.sin(alpha)*np.sin(gamma)*np.cos(phi_0)
            + np.cos(gamma)*np.sin(phi_0)*np.sin(theta_0))
```

**Comment by wmwv** [comment]
And in `healpix_geom.c`

```
    f1vec.x = cos_alpha * sin_gamma * cos_phi * cos_theta + sin_alpha * sin_gamma * sin_phi +
              cos_gamma * cos_phi * sin_theta;
    f2vec.x = -cos_alpha * sin_gamma * cos_phi * cos_theta - sin_alpha * sin_gamma * sin_phi +
              cos_gamma * cos_phi * sin_theta;
    f1vec.y = cos_alpha * sin_gamma * sin_phi * cos_theta - sin_alpha * sin_gamma * cos_phi +
              cos_gamma * sin_phi * sin_theta;
    f2vec.y = -cos_alpha * sin_gamma * sin_phi * cos_theta + sin_alpha * sin_gamma * cos_phi +
              cos_gamma * sin_phi * sin_theta;
```

File: hpgeom/healpix_geom.c (hunk 6 above)
Comment (erykoff) [reply hpgeom/healpix_geom.c]: I also fixed the docstring which was similarly incorrect/unclear.

**Comment by erykoff** [comment]
Great catch on that sign error!  Yes, I had already flipped the sign without making a comment in the code.  I have now fixed the sign error and also added a comment to the code that alpha is opposite what it was in the stack overflow example so that future us will be able to follow better.

I checked my example script for all sorts of rotation angles and they all look good now.

**Review by wmwv — APPROVED** [review]
LGTM

**Pull request merged**